### PR TITLE
Update Poutine docs

### LINKED ISSUE #826: Improve documentation of some Pyro internals
This issue tracks some documentation improvements that should be made before the 0.2 release.
- [x] Improve documentation of `Marginal` or remove it (#543)
- [x] Improve documentation of `TracePosterior` or remove it
- [x] Improve documentation of `Trace` 
- [ ] Correct and improve introductory documentation for Poutine
- [x] Format everything correctly for sphinx: handlers and `Trace` at top level, `Messenger`s and `runtime` collapsed

### PR DESCRIPTION
**Pull request by eb8680**
Resolves #826 

This PR updates the docs of the `pyro.poutine` module to be more descriptive and to reflect all the changes that have happened.

### CODE REVIEW

**Comment by eb8680** [comment]
I won't change any code in this PR, only Sphinx and docstrings.  That way we don't need to run the full test suite to merge.

File: pyro/poutine/handlers.py [hunk 1]
```diff
@@ -25,71 +25,132 @@
...
+        ...     return z ** 2
+
+    We can record its execution using ``trace``
+    and use the resulting data structure to compute the log-joint probability
+    of all of the sample sites in the execution or extract all parameters.
 
-    Adds trace data structure site constructors to primitive stacks
+        >>> trace = pyro.poutine.trace(model).get_trace(0.0)
+        >>> logp = trace.log_prob_sum()
+        >>> params = [trace.nodes[name]["value"].unconstrained() for name in trace.param_nodes]
```
Comment (fritzo) [inline pyro/poutine/handlers.py]: Which of `["value"]` or `["value"].unconstrained()` do you think is the more common use case?

Comment (eb8680) [reply pyro/poutine/handlers.py]: The latter is always guaranteed to be correct, so I used it

File: pyro/poutine/handlers.py [hunk 2]
```diff
@@ -165,7 +234,7 @@ def scale(fn=None, scale=None):
 
 def indep(fn=None, name=None, size=None, dim=None):
     """
-    Alias for IndepMessenger constructor.
+    Note: Low-level; use `pyro.iarange` instead.
```
Comment (jpchen) [inline pyro/poutine/handlers.py]: you can turn this into a highlighted note block with `.. note::`

File: pyro/poutine/block_messenger.py [hunk 3]
```diff
@@ -4,9 +4,20 @@
...
+    Default behavior: block everything
+    BlockMessenger has a flexible interface that allows users
+    to specify in several different ways
+    which sites should be hidden or exposed.
+
+    A site is hidden if at least one of the following holds:
+
+        1. msg["name"] in hide
+        2. msg["type"] in hide_types
+        3. msg["name"] not in expose and msg["type"] not in expose_types
```
Comment (jpchen) [inline pyro/poutine/block_messenger.py]: does this render correctly?

File: pyro/poutine/handlers.py [hunk 4]
```diff
@@ -25,71 +25,132 @@
...
+    A site is hidden if at least one of the following holds:
+
+        1. ``msg["name"] in hide``
+        2. ``msg["type"] in hide_types``
+        3. ``msg["name"] not in expose and msg["type"] not in expose_types``
+        4. ``hide``, ``hide_types``, and ``expose_types`` are all ``None``
+
+    For example, suppose the stochastic function fn has two sample sites "a" and "b".
+    Then any effect outside of ``BlockMessenger(fn, hide=["a"])``
+    will not be applied to site "a" and will only see site "b":
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/poutine/handlers.py]: I guess not. Deleting comment.

File: pyro/poutine/handlers.py [hunk 5]
```diff
@@ -25,71 +25,131 @@
...
     at those sites in the new trace
+
+    Consider the following Pyro program:
+
+        >>> def model(x):
+        ...     s = pyro.param("s", torch.tensor(0.5))
+        ...     z = pyro.sample("z", dist.Normal(x, s))
+        ...     return z ** 2
+
+    ``replay`` makes ``sample`` statements behave as if they had sampled the values
```
Comment (fritzo) [inline pyro/poutine/handlers.py]: Maybe emphasize "they had sampled the recorded values"?

File: pyro/poutine/block_messenger.py (hunk 3 above)
Comment (fritzo) [reply pyro/poutine/block_messenger.py]: Indeed it does.

File: pyro/poutine/handlers.py [hunk 6]
```diff
@@ -25,71 +25,131 @@
...
+        >>> tr = trace(lifted_model).get_trace(0.0)
+        >>> tr.nodes["s"]["type"] == "sample"
+        True
+        >>> tr2 = trace(lifted_model).get_trace(0.0)
+        >>> (tr2.nodes["s"]["value"] == tr.nodes["s"]["value"]).all()
+        False
+
+    :param fn: function whose parameters will be lifted to random values
+    :param prior: prior function in the form of a Distribution or a dict of stochastic fns
+    :returns: ``fn`` decorated with a :class:`LiftMessenger`
```
Comment (fritzo) [inline pyro/poutine/handlers.py]: For this to link correctly, I believe you need to use
```
:class:`~pyro.poutine.messengers.LiftMessenger`
```
and similar for other `Messenger` references in this file.

**Comment by fritzo** [comment]
Ready to merge?

**Pull request merged**